### PR TITLE
object: unique username for OBC even when preceding OBC was retained

### LINKED ISSUE #11345: ObjectBucketClaim user can conflict with previously created users
When creating an ObjectBucketClaim with name `name` in namespace `namespace`, Rook will create a user called `obc-namespace-name`. This username might conflict with existing users, e.g.

- if an ObjectBucketClaim with the same name in the same namespace once existed and was deleted or
- ~if an ObjectBucketClaim with the same name in the same namespace exists in a different K8s cluster (multi-site).~ (fixed in #11665)

In particular, in the first case, if the old ObjectBucketClaim was deleted with `reclaimPolicy: Retain`, Rook will fail to create the new bucket with the following error, because the user already owns a bucket.

```
rook-ceph-operator-56f946595-djgv7 rook-ceph-operator 2022-11-23 15:20:36.672025 I | op-bucket-prov: Provision: creating bucket "test-bucket-626a4eb9-6fb3-414b-854e-059619c2cb38" for OBC "test-bucket"
rook-ceph-operator-56f946595-djgv7 rook-ceph-operator 2022-11-23 15:20:36.672072 I | op-bucket-prov: creating Ceph user "obc-test-test-bucket"
rook-ceph-operator-56f946595-djgv7 rook-ceph-operator 2022-11-23 15:20:36.683362 I | op-bucket-prov: successfully created Ceph user "obc-test-test-bucket" with access keys
rook-ceph-operator-56f946595-djgv7 rook-ceph-operator 2022-11-23 15:20:36.687323 I | ceph-object-controller: creating bucket "test-bucket-626a4eb9-6fb3-414b-854e-059619c2cb38"
rook-ceph-operator-56f946595-djgv7 rook-ceph-operator E1123 15:20:36.691341       1 controller.go:204] error syncing 'test/test-bucket': error provisioning bucket: error creating bucket "test-bucket-626a4eb9-6fb3-414b-854e-059619c2cb38": failed to create bucket "test-bucket-626a4eb9-6fb3-414b-854e-059619c2cb38": TooManyBuckets: 
```

Proposed solution: Rook should append a random unique identifier to generate the username and track the generated username in the ObjectBucketClaim spec/status.

**Environment**:
* Rook version: v1.10.6
* External Ceph cluster, Rook has `client.admin` credentials.

### PR DESCRIPTION
**Pull request by haslersn**
**Description of your changes:**

For an OBC's name we cannot simply use the OBC's namespace and name, because they can be reused, while the preceding bucket might be retained by reclaimPolicy. (PR #11665 also didn't solve this issue, as it is an intra-cluster issue.) Therefore we instead use the OBC's UID which should be unique within the cluster and across clusters.

Similar to how it was after #11665, existing users will continue to work, i.e., rook will not update the `ObjectBucket.spec.additionalState.cephUser` field of existing buckets.

**Which issue is resolved by this Pull Request:**
Resolves #11345

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by BlaineEXE** [comment]
I think this is working to address the same class of bug as this other PR https://github.com/rook/rook/pull/12880#discussion_r1321752613.

I believe that this implementation is a simpler approach that I would prefer.

One requirement, @haslersn : please copy the commit description to this PR itself for easier discussion within GitHub.

File: pkg/operator/ceph/object/bucket/rgw-handlers.go [hunk 1]
```diff
@@ -47,15 +48,11 @@ func (p *Provisioner) createCephUser(username string) (accKey string, secKey str
...
-	fsid := p.clusterInfo.FSID
-	if p.clusterInfo.FSID == "" {
-		return "", errors.Errorf("failed to find ceph cluster FSID")
-	}
-	return "obc-" + obcNamespace + "-" + obcName + "-" + fsid, nil
+func (p *Provisioner) genUserName(obc *v1alpha1.ObjectBucketClaim) string {
+	// user name can be deterministically generated from the OBC's UID. We
+	// cannot simply use the OBC's namespace and name, because they can be
+	// reused, while the preceding bucket might be retained by reclaimPolicy.
+	return "obc-" + string(obc.UID)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/bucket/rgw-handlers.go:55]: Personally, I like the idea of keeping `obcNamespace` and `obcName` here to help with visual inspection of users on the backend and mapping them into the k8s domain. This isn't a hard requirement, but I believe it's a worthwhile "hint" that can help when doing complex debugging, and it's essentially free.

```suggestion
	return "obc-" + obc.Namespace + "-" obc.Name + string(obc.UID)
```

**Review by BlaineEXE — APPROVED** [review]
https://github.com/rook/rook/pull/12884#discussion_r1322070091

I think this can be disregarded. While namespace names are limited to 63 chars, the k8s spec says that resource names can be up to 253, which is approaching the known 255 limit for obc "friendly" user names. While I seem to recall being limited to 63 chars for CR names, I may be mistaken, and it seems safest to proceed with this PR as-is.

**Comment by BlaineEXE** [comment]
Considering point 2 of Travis' question here: https://github.com/rook/rook/pull/12880#discussion_r1321782253

I think it would be good to create an OBC before this change and then ensure it still works afterwards. I thought we checked this in Rook's upgrade test, but I am seeing that isn't the case.

**Comment by BlaineEXE** [comment]
> Considering point 2 of Travis' question here: [#12880 (comment)](https://github.com/rook/rook/pull/12880#discussion_r1321782253)
> 
> I think it would be good to create an OBC before this change and then ensure it still works afterwards. I thought we checked this in Rook's upgrade test, but I am seeing that isn't the case.

Related Issue describing the need for OBC testing during upgrade CI: https://github.com/rook/rook/issues/12887

**Comment by haslersn** [comment]
I tested this PR by cherry-picking 041fc1a71ce5fc6c6f7be323fb855642a7c86452 on top of v1.12.3, building and deploying that. I confirm that now deleting and recreating an ObjectBucketClaim with the same namespace and name works, even if the previous bucket is retained.

**Comment by BlaineEXE** [comment]
@travisn thoughts on 1.11 bp here?

**Comment by travisn** [comment]
> @travisn thoughts on 1.11 bp here?

Sounds good to could include. In general, if someone asks for a 1.11 backport we can still do it if needed.

**Pull request merged**